### DISCUSSION TITLE
High Availability/Load Balancing

### DISCUSSION THREAD

**Discussion by IvanRibakov** [sipsorcery-org/sipsorcery#1147] [Q&A] [answered]
Are there any examples/known implementations that used SIPSorcery to achieve Load Balancing/HA? Any known issues or limitations?

**Comment by sipsorcery** [comment]
The library was originally written in 2005/2006 to act as a Stateless SIP Proxy for a commercial VoIP service. It's no longer used on that service, and in fact that company has now shut down.

As far as HA there are no specific features in the library to support it but that will be the casse for most SIP Proxy implementations. HA is typically achieved with SRV records used to load balance across multiple proxy instances.

**Reply by IvanRibakov** [reply]
@sipsorcery Thanks for getting back on this. I understand that there is always an option to distribute the load using SRV records. I was more looking into mechanisms to ensure service continuity during the fallback scenario (active/stand-by pattern). Typically this requires a means to share the dialog and transaction states between cluster instances. 
Another common challenge is ensuring consistent routing towards the serving instance (in case of active/active pattern). 
Do you know if SIPSorcery was used in any of the above scenarios?

**Reply by sipsorcery** [reply, marked as answer]
No, not that I know of. 

That sort of capabiltity is the remit of the sophisticated SIP Proxy apps like SER & Kamilio. Typically you'd be looking at a DB or some kind of persistent store to maintain the high velocity state. In production SIP Proxies I've only ever used this library in stateless SIP Proxy mode.
